### PR TITLE
Conversions

### PR DESCRIPTION
**Pull request by prakharcode**
This PR resolves conversion for all the functions required in AstroBase.
@helgee there might be a need to relocate some functions in different files. There are approximation errors in 4 functions.

A comprehensive report for all the conversion can be found [here](https://docs.google.com/spreadsheets/d/1sDMmNfbCe3PSAPTwOXko5ae7FnsHnSTPrgIRv-513PY/edit?usp=sharing)

### CODE REVIEW

File: src/EarthAttitude.jl [hunk 1]
```diff
@@ -1007,4 +1042,654 @@ function nutation_00a(jd1, jd2)
...
+ -0.00165602  3.35578e-5   0.999999
+```
+"""
+function precession_nutation_matrix_a06(jd1, jd2)
+    gamb, phib, psib, epsa = precession_fukushima_williams06(jd1, jd2)
+    dp, de = nutation_a06(jd1, jd2)
+    fukushima_williams_matrix(gamb, phib, psib + dp, epsa + de)
+end
+"""
+approx error
```
Comment (helgee) [inline src/EarthAttitude.jl]: What's the issue here?

File: src/EarthAttitude.jl [hunk 2]
```diff
@@ -1007,4 +1042,654 @@ function nutation_00a(jd1, jd2)
...
+"""
+function s00a(jd1, jd2)
+    rbpn = precession_nutation_matrix_a00(jd1, jd2)
+    s00(jd1, jd2, rbpn[1, 3], rbpn[2, 3])
+end
+
+"""
+approx error
```
Comment (helgee) [inline src/EarthAttitude.jl]: And here?

File: src/EarthAttitude.jl [hunk 3]
```diff
@@ -1007,4 +1042,654 @@ function nutation_00a(jd1, jd2)
...
+"""
+function nutation_matrix_day(jd1, jd2)
+    eps = mean_obliquity_of_ecliptic(jd1, jd2)
+    dp, de = nutation_a06(jd1, jd2)
+
+    numat(eps, dp, de)
+end
+
+"""
+approx error
```
Comment (helgee) [inline src/EarthAttitude.jl]: And here?

File: src/EarthAttitude.jl [hunk 4]
```diff
@@ -1007,4 +1042,654 @@ function nutation_00a(jd1, jd2)
...
+"""
+function xys00a(jd1, jd2)
+    rbpn = precession_nutation_matrix_a00(jd1, jd2)
+    rbpn[1,3], rbpn[2,3], s00(jd1, jd2, rbpn[1,3], rbpn[2,3])
+end
+
+"""
+approx error
```
Comment (helgee) [inline src/EarthAttitude.jl]: And here? 😂

**Review by helgee — COMMENTED** [review]
What do you mean by "approximation errors" @prakharcode?

File: src/EarthAttitude.jl (hunk 1 above)
Comment (prakharcode) [reply src/EarthAttitude.jl]:
```
julia> s00a(2.4578265e6, 0.30434616919175345)
2.9183401976144617e-8

julia> ERFA.s00a(2.4578265e6, 0.30434616919175345)
2.9177473009711407e-8
```

File: src/EarthAttitude.jl (hunk 2 above)
Comment (prakharcode) [reply src/EarthAttitude.jl]:
```
julia> s00b(2.4578265e6, 0.30434616919175345)
2.9182068810695328e-8

julia> ERFA.s00b(2.4578265e6, 0.30434616919175345)
2.917613983822591e-8
```

File: src/EarthAttitude.jl (hunk 4 above)
Comment (prakharcode) [reply src/EarthAttitude.jl]:
```

julia> xys00b(2.4578265e6, 0.30434616919175345)
(0.001655885521808701, -3.987456146931851e-5, 2.9182068810695328e-8)

julia> ERFA.xys00b(2.4578265e6, 0.30434616919175345)
(0.0016558855494194407, -3.986739971767614e-5, 2.917613983822591e-8)
```

File: src/EarthAttitude.jl (hunk 3 above)
Comment (prakharcode) [reply src/EarthAttitude.jl]:
```
julia> xys00a(2.4578265e6, 0.30434616919175345)
(0.001655885227051451, -3.987617878202287e-5, 2.9183401976144617e-8)

julia> ERFA.xys00a(2.4578265e6, 0.30434616919175345)
(0.0016558852546651388, -3.9869017036298323e-5, 2.9177473009711407e-8)
```

**Comment by helgee** [comment]
These deviations are far too large to be "approximation errors". Something must be wrong.

Please test the return values of these functions individually so the tests actually fail, e.g.

```julia
let
    x1, y1, s1 = xys00b(2.4578265e6, 0.30434616919175345)
    x2, y2, s2 = ERFA.xys00b(2.4578265e6, 0.30434616919175345)
    @test x1 ≈ x2
    @test y1 ≈ y2
    @test s1 ≈ s2
end
```

**Comment by helgee** [comment]
I think we have a general problem here because right now we do many comparisons of the form
```julia
@test a::AbstractArray ≈ b::AbstractArray
```
which probably hides many inaccuracies. The code above does a norm-based comparison.

We should probably rewrite all array tests to one of these forms:

```julia
# Has the disadvantage of not showing which element failed
@test all(a .≈ b)

# Has the disadvantage of being cluncky
for i in eachindex(a)
    @eval @test a[$i] ≈ b[$i]
end
```

What do you think @giordano?

**Comment by helgee** [comment]
@prakharcode Since the coding period is almost over, I would suggest that you fix the missing docstrings (the "approx errors"), I merge this, and I will tackle the accuracy problems in a separate PR.

**Comment by prakharcode** [comment]
Yes, I was actually doing the `for loop` for each one there are some errors which gets carried out throughout the code.

`bp00` in particular.

**Comment by prakharcode** [comment]
Okay, I'll fix the docstrings :smile:

**Comment by giordano** [comment]
I honestly don't know how this kind of tests is carried out elsewhere, e.g. in julia itself. Maybe the for loop is a good solution for the time being

**Comment by helgee** [comment]
I had a look at Base and the solution is actually quite simple:

```julia
julia> @testset for i in eachindex(a)
           @test a[i] ≈ b[i]
       end
i = 1: Test Failed at REPL[21]:2
  Expression: a[i] ≈ b[i]
   Evaluated: -1.4339454881257854 ≈ 0.2938678671127828
Stacktrace:
 [1] macro expansion at ./REPL[21]:2 [inlined]
 [2] top-level scope at /Users/helge/projects/julia/dev/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1152 [inlined]
 [3] top-level scope at ./none:0
 [4] eval(::Module, ::Any) at ./boot.jl:319
 [5] eval_user_input(::Any, ::REPL.REPLBackend) at /Users/helge/projects/julia/dev/usr/share/julia/stdlib/v1.0/REPL/src/REPL.jl:85
 [6] macro expansion at /Users/helge/projects/julia/dev/usr/share/julia/stdlib/v1.0/REPL/src/REPL.jl:117 [inlined]
 [7] (::getfield(REPL, Symbol("##28#29")){REPL.REPLBackend})() at ./task.jl:262
Test Summary: | Fail  Total
i = 1         |    1      1
Test Summary: | Fail  Total
i = 1         |    1      1
ERROR: Some tests did not pass: 0 passed, 1 failed, 0 errored, 0 broken.
```

**Pull request merged**